iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -307,6 +307,318 @@ class AdjustDepthwiseFilterShape : public OpRewritePattern<mhlo::ConvOp> {
+    int64_t rhsContractionDim = (*op.dot_dimension_numbers()
+                                      .rhs_contracting_dimensions()
+                                      .int_value_begin())
+                                    .getLimitedValue();
+    if (lhsContractionDim == 2 && rhsContractionDim == 1) return failure();
+
+    Value lhs = op.lhs(), rhs = op.rhs();
+
+    auto make1DElementsAttr = [&rewriter](ArrayRef<int64_t> integers) {
+      auto type = RankedTensorType::get({static_cast<int64_t>(integers.size())},
+                                        rewriter.getIntegerType(64));
+      return DenseIntElementsAttr::get(type, integers);
+    };
+
+    // Transpose lhs input.
+    if (lhsContractionDim == 1) {
+      Type transposedType = RankedTensorType::get(
+          {lhsShapeType.getDimSize(0), lhsShapeType.getDimSize(2),
+           lhsShapeType.getDimSize(1)},
+          resultType.getElementType());
+      lhs = rewriter.create<mhlo::TransposeOp>(op.getLoc(), transposedType, lhs,
+                                               make1DElementsAttr({0, 2, 1}));
+    }
+
+    if (rhsContractionDim == 2) {
+      Type transposedType = RankedTensorType::get(
+          {rhsShapeType.getDimSize(0), rhsShapeType.getDimSize(2),
+           rhsShapeType.getDimSize(1)},
+          resultType.getElementType());
+      rhs = rewriter.create<mhlo::TransposeOp>(op.getLoc(), transposedType, rhs,
+                                               make1DElementsAttr({0, 2, 1}));
+    }
+
+    auto dimensionNumbers = mhlo::DotDimensionNumbers::get(
+        /*lhs_batching_dimensions=*/make1DElementsAttr({0}),
+        /*rhs_batching_dimensions=*/make1DElementsAttr({0}),
+        /*lhs_contracting_dimensions=*/make1DElementsAttr({2}),
+        /*rhs_contracting_dimensions=*/
+        make1DElementsAttr({1}), rewriter.getContext());
+
+    Value result = rewriter.create<mhlo::DotGeneralOp>(
+        op.getLoc(), op.getType(), lhs, rhs, dimensionNumbers,
+        op.precision_config().hasValue() ? op.precision_config().getValue()
+                                         : nullptr);
+    rewriter.replaceOp(op, result);
+    return success();
+  }
+};
+
+// Rewrite mhlo.dot_general to operate on rank-3 tensors when reduction dims are
+// in consequative order and not spliting the domain.
+class RankReducedDotGeneral : public OpRewritePattern<mhlo::DotGeneralOp> {
+ public:
+  using OpRewritePattern<mhlo::DotGeneralOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::DotGeneralOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhsShapeType = op.lhs().getType().cast<ShapedType>();
+    auto rhsShapeType = op.rhs().getType().cast<ShapedType>();
+    auto resultType = op.getResult().getType().cast<ShapedType>();
+
+    if (!lhsShapeType || !rhsShapeType || !resultType) return failure();
+    if (!lhsShapeType.hasStaticShape() || !rhsShapeType.hasStaticShape())
```

Comment:
Here too, does it need to be static shape?

Reply:
Yes, this one needs to be static to perform the reshape.

Reply:
How about moving this to the lowering of dot_general to linalg.batch_matmul. There we could use a `linalg.tensor_reshape` and handle this for dynamic shapes as well. Most of the logic should hold for that.

Reply:
This will need to use mhlo::DynamicReshapeOp instead, and we will need to add a pattern for lowering it to tensor_reshape.

Actually, I don't really know how would we lower it to tensor_reshape? How do we construct reassociation maps? Do we just map first n dims to another first dim, and map 1:1 for remaining, like [(d0, d1, d2) -> (d0), (d3) -> (d1), (d4) -> (d2)]?

Anyway, I think it's fine to have static shape check here, and file an issue for tracking dynamic shape.

Reply:
Using DynamicReshapeOp is fine. What you suggest is exactly what I had in mind. The reassociation map is a bit different though

For the output it would be :

````
[affine_map<(d0, d1, d2, ...,dn, dm, dk) -> (d0, d1, d2, ... dn)>,
 affine_map<(d0, d1, d2, ...,dn, dm, dk) -> (dm)>,
 affine_map<(d0, d1, d2, ...,dn, dm, dk) -> (dk)>]
```

If there is any other issue we can just add this for now and file and issue for dynamic shape case, but it seems like a fairly straight-forward thing to do right now. I would rather spend a day or so (at the most) to just do it for dynamic shape instead of doing static shape now and then revisiting this. At the very least if there is an issue for the dynamic shape case we would know about it now.

Reply:
Wasn't aware of DynamicReshapeOp, thanks for the pointer. I agree lets emit DynamicReshapeOp when possible.

Reply:
I am not aware of dynamic reshape op. Can it be lowered to a `linalg.tensor_reshape`. If not I would suggest doing this pattern in the MHLO -> Linalg on Buffers conversion

Reply:
If we moved this to linalg on buffers we will also need in some cases a transpose on buffers inserted in the same region (vs here where its outside and in tensors) which will change the memory access of the `*_matmul`. I think a static dim reduction HLO->HLO pattern is useful when we hit illegal dot_general with dim dims we can write a linalg conversion.

Reply:
Isnt that better. We wont be materializing a tensor just to be a result of the transpose. Even if a transpose is needed we can insert that here. Basically do everything apart from the "reshape" so that it gets to the iteration space to form of [<outer_batch_dims>, M, N, K]. I still don't see a fundamental reason for this to be done just for static shapes.

Reply:
I see your point now. Sure, we can extend the transpose pattern here to work on any ranked tensor not only three and move the rank-reduction part to linalg on buffers.

Reply:
Ok, so I am assuming you are going to do this as a follow up task. Could you please file an issue to track this if you want to do this as a follow up change?